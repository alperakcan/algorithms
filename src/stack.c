
#include <stdlib.h>

#include "list.h"
#include "stack.h"

struct item {
	struct list list;
	void *data;
};

struct stack {
	struct list list;
	unsigned int nitems;
};

struct stack * stack_create (void)
{
	int rc;
	struct stack *stack;
	stack = malloc(sizeof(*stack));
	if (stack == NULL) {
		return NULL;
	}
	rc = list_init(&stack->list);
	if (rc != 0) {
		free(stack);
		return NULL;
	}
	stack->nitems = 0;
	return stack;
}

static int stack_destroy_iterator (void *context, struct list *list, struct list *item)
{
	struct item *i;
	(void) context;
	i = list_containerof(item, struct item, list);
	list_del(list, &i->list);
	free(i);
	return 0;
}

int stack_destroy (struct stack *stack)
{
	if (stack == NULL) {
		return -1;
	}
	list_iterate_safe(&stack->list, stack_destroy_iterator, NULL);
	free(stack);
	return 0;
}

int stack_push (struct stack *stack, void *data)
{
	struct item *item;
	if (stack == NULL) {
		return -1;
	}
	item = malloc(sizeof(*item));
	if (item == NULL) {
		return -1;
	}
	item->data = data;
	list_add(&stack->list, &item->list);
	stack->nitems += 1;
	return 0;
}

void * stack_pop (struct stack *stack)
{
	void *data;
	struct list *l;
	struct item *i;
	if (stack == NULL) {
		return NULL;
	}
	if (stack->nitems <= 0) {
		return NULL;
	}
	l = list_first(&stack->list);
	if (l == NULL) {
		return NULL;
	}
	i = list_containerof(l, struct item, list);
	if (i == NULL) {
		return NULL;
	}
	data = i->data;
	list_del(&stack->list, &i->list);
	free(i);
	stack->nitems -= 1;
	return data;
}

unsigned int stack_size (struct stack *stack)
{
	if (stack == NULL) {
		return 0;
	}
	return stack->nitems;
}

struct stack_iterator_context {
	struct stack *stack;
	int (*iterator) (void *context, struct stack *stack, void *item);
	void *context;
};

static int stack_iterator (void *context, struct list *list, struct list *item)
{
	struct item *i;
	struct stack_iterator_context *c;
	(void) list;
	c = context;
	i = list_containerof(item, struct item, list);
	c->iterator(c->context, c->stack, i->data);
	return 0;
}

int stack_iterate (struct stack *stack, int (*iterator) (void *context, struct stack *stack, void *item), void *context)
{
	struct stack_iterator_context c;
	if (stack == NULL) {
		return -1;
	}
	if (iterator == NULL) {
		return -1;
	}
	c.iterator = iterator;
	c.context = context;
	c.stack = stack;
	list_iterate(&stack->list, stack_iterator, &c);
	return 0;
}
