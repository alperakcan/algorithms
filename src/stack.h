
struct stack;

struct stack * stack_create (void);
int stack_destroy (struct stack *stack);
int stack_push (struct stack *stack, void *data);
void * stack_pop (struct stack *stack);
unsigned int stack_size (struct stack *stack);
int stack_iterate (struct stack *stack, int (*iterator) (void *context, struct stack *stack, void *item), void *context);
