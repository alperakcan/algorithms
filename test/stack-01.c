
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "stack.h"

static int iterator (void *context, struct stack *stack, void *data)
{
	(void) context;
	(void) stack;
	printf("%s ", (char *) data);
	return 0;
}

int main (int argc, char *argv[])
{
	int i;
	struct stack *stack;
	stack = stack_create();
	if (stack == NULL) {
		return -1;
	}
	for (i = 1; i < argc; i++) {
		stack_push(stack, argv[i]);
	}
	stack_iterate(stack, iterator, NULL);
	stack_destroy(stack);
	return 0;
}
