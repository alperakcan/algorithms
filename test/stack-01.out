is - - - that - - be - to not or be to 